pkg/networkengine/vpndriver/libreswan/libreswan.go

```diff
@@ -0,0 +1,275 @@
+package libreswan
+
+import (
+	"fmt"
+	"os"
+	"os/exec"
+	"sort"
+	"syscall"
+	"time"
+
+	"github.com/vdobler/ht/errorlist"
+	"k8s.io/klog/v2"
+
+	"github.com/openyurtio/raven/cmd/agent/app/config"
+	netlinkutil "github.com/openyurtio/raven/pkg/networkengine/util/netlink"
+	"github.com/openyurtio/raven/pkg/networkengine/vpndriver"
+	"github.com/openyurtio/raven/pkg/types"
+)
+
+const DriverName = "libreswan"
+
+var _ vpndriver.Driver = (*libreswan)(nil)
+
+func init() {
+	vpndriver.RegisterDriver(DriverName, New)
+}
+
+const (
+	SecretFile string = "/etc/ipsec.d/raven.secrets"
+	DefaultPSK string = "openyurt-raven"
+)
+
+type libreswan struct {
+	connections map[string]struct{}
+	nodeName    types.NodeName
+}
+
+func (l *libreswan) Init() error {
+	// Ensure secrets file
+	_, err := os.Stat(SecretFile)
+	if err == nil {
+		os.Remove(SecretFile)
+	}
+	file, err := os.Create(SecretFile)
+	if err != nil {
+		klog.Errorf("fail to create secrets file: %v", err)
+	}
+	defer file.Close()
+
+	psk := os.Getenv("LIBRESWAN_PSK")
+	if psk == "" {
+		psk = DefaultPSK
+		klog.Warning(fmt.Sprintf("use weak PSK: %s", psk))
+	}
+	fmt.Fprintf(file, "%%any %%any : PSK \"%s\"\n", psk)
+
+	return l.runPluto()
+}
+
+func New(cfg *config.Config) (vpndriver.Driver, error) {
+	return &libreswan{
+		connections: map[string]struct{}{},
+		nodeName:    types.NodeName(cfg.NodeName),
+	}, nil
+}
+
+func (l *libreswan) Apply(network *types.Network) error {
+	errList := errorlist.List{}
+	if network.LocalEndpoint == nil || len(network.RemoteEndpoints) == 0 {
+		klog.Info("no local gateway or remote gateway is found, cleaning vpn connections")
+		return l.Cleanup()
+	}
+	if network.LocalEndpoint.NodeName != l.nodeName {
+		klog.Infof("the current node is not gateway node, cleaning vpn connections")
+		return l.Cleanup()
+	}
+	centralGw := l.findCentralGw(network)
+	leftSubnets := make([]string, 0)
+	leftSubnets = append(leftSubnets, network.LocalEndpoint.Subnets...)
+	// In order to tell the central gateway to forward traffic between gateways behind NAT device,
+	// we should put subnets of those gateways into left subnets.
+	if centralGw != nil && network.LocalEndpoint.GatewayName == centralGw.GatewayName {
+		for _, remoteGw := range network.RemoteEndpoints {
+			if remoteGw.UnderNAT {
+				leftSubnets = append(leftSubnets, remoteGw.Subnets...)
```

Comment:
@njucjc Thanks for your careful review, you are right, the central gateway will set up unexpected connection [B-B].

But I think the subnets of remote gateway are still needed in `leftsubnet`. 
In your case, the complete vpn setup for `cental` gateway should be:
1. The direct vpn connections to `gw1` and `gw2`: [A-B], [A-C] 
2. The connections used to forward traffic for `gw1` and `gw2`: [B-C], [C-B]

Both B and C are in  left subnet.

So I think we just need to continue the loop when `leftSubnet` == `rightSubnet`, like this:
```
	for _, leftSubnet := range leftSubnets {
		for _, remoteGw := range network.RemoteEndpoints {
			for _, rightSubnet := range remoteGw.Subnets {
				if rightSubnet == leftSubnet { // < -- add this statement
					continue
				}
....
```

Reply:
> @njucjc Thanks for your careful review, you are right, the central gateway will set up unexpected connection [B-B].
> 
> But I think the subnets of remote gateway are still needed in `leftsubnet`. In your case, the complete vpn setup for `cental` gateway should be:
> 
> 1. The direct vpn connections to `gw1` and `gw2`: [A-B], [A-C]
> 2. The connections used to forward traffic for `gw1` and `gw2`: [B-C], [C-B]
> 
> Both B and C are in left subnet.
> 
> So I think we just need to continue the loop when `leftSubnet` == `rightSubnet`, like this:
> 
> ```
> 	for _, leftSubnet := range leftSubnets {
> 		for _, remoteGw := range network.RemoteEndpoints {
> 			for _, rightSubnet := range remoteGw.Subnets {
> 				if rightSubnet == leftSubnet { // < -- add this statement
> 					continue
> 				}
> ....
> ```


This does not completely solve the problem, assuming that `gw3` contains more than one subnets, such as [D, E], in this case, `leftsubnet` will be [A, B, C, D, E], when central gateway trys to set up vpn connection to the gw3 (subnets: [D, E]), it will set up  vpn connections (A-D, A-E, ..., D-E, E-D), which the D-E and E-D connections are unexpected.

Reply:
> @njucjc Thanks for your careful review, you are right, the central gateway will set up unexpected connection [B-B].
> 
> But I think the subnets of remote gateway are still needed in `leftsubnet`. In your case, the complete vpn setup for `cental` gateway should be:
> 
> 1. The direct vpn connections to `gw1` and `gw2`: [A-B], [A-C]
> 2. The connections used to forward traffic for `gw1` and `gw2`: [B-C], [C-B]
> 
> Both B and C are in left subnet.
> 
> So I think we just need to continue the loop when `leftSubnet` == `rightSubnet`, like this:
> 
> ```
> 	for _, leftSubnet := range leftSubnets {
> 		for _, remoteGw := range network.RemoteEndpoints {
> 			for _, rightSubnet := range remoteGw.Subnets {
> 				if rightSubnet == leftSubnet { // < -- add this statement
> 					continue
> 				}
> ....
> ```

what we need should be:

1. `leftSubnets` == [A, C], when `central` gateway connects to `gw1`, setting up vpn connections A-B and C-B
2. `leftSubnets` == [A, B], when `central` gateway connects to `gw2`, setting up vpn connections A-C and B-C

Reply:
@njucjc Hi, I've updated and added some test cases for it, please take a look.
https://github.com/openyurtio/raven/compare/a18732eb5db7db4875b98d7e7f0b6302db044492..a2c3d49c06294a11081b28e43b5839848f4d1e51

Reply:
> 

LGTM